webots/controllers/RL_Supervisor/agent.py

```diff
@@ -0,0 +1,544 @@
+
+            self.data_sent = self.__serialmux.send_data(
+                "SPEED_SET", motorcontrol
+            )  # stop the motors immediately
+
+            # Failed to send data. Appends the data to unsent_data List
+            if self.data_sent is False:
+                self.unsent_data.append(("SPEED_SET", motorcontrol))
+
+            self.data_sent = self.__serialmux.send_data("CMD", cmd_payload)
+
+            # Failed to send data. Appends the data to unsent_data List
+            if self.data_sent is False:
+                self.unsent_data.append(("CMD", cmd_payload))
+
+    def normalize_sensor_data(self, sensor_data):
+        """The normalize_sensor_data function scales the sensor data to a range between 0 and 1."""
+
+        sensor_data = np.array(sensor_data) / MAX_SENSOR_VALUE
+        return sensor_data
+
+    def calculate_position(self, sensor_data):
+        """
+        Determines the deviation and returns an estimated position of the robot
+        with respect to a line. The estimate is made using a weighted average of
+        the sensor indices multiplied by 1000, so that a return value of 0
+        indicates that the line is directly below sensor 0, a return value of
+        1000 indicates that the line is directly below sensor 1, 2000
+        indicates that it's below sensor 2000, etc.  Intermediate values
+        indicate that the line is between two sensors. The formula is:
+
+                0*value0 + 1000*value1 + 2000*value2 + ...
+            --------------------------------------------
+                value0  +  value1  +  value2 + ...
+
+        This function assumes a dark line (high values) surrounded by white
+        (low values).
+
+        Parameters
+        ----------
+        sensor_data : List
+
+        Returns
+        ----------
+        Estimated position with respect to track.
+        """
+
+        estimated_pos = 0
+        numerator = 0
+        denominator = 0
+        weight = 1000
+        is_on_line = False
+
+        max_sensors = len(sensor_data)
+        sensor_max_value = 1000
+        last_pos_value = 0
+
+        for idx, sensor_value in enumerate(sensor_data):
+            numerator += idx * weight * sensor_value
+            denominator += sensor_value
+
+            # Keep track of whether we see the line at all.
+            if LINE_SENSOR_ON_TRACK_MIN_VALUE < sensor_value:
+                is_on_line = True
+
+        if False is is_on_line:
+
+            # If it last read to the left of center, return 0.
+            if last_pos_value < ((max_sensors - 1) * sensor_max_value) / 2:
+                estimated_pos = 0
+
+            # If it last read to the right of center, return the max. value.
+            else:
+                estimated_pos = (max_sensors - 1) * sensor_max_value
+
+        else:
+            # Check to avoid division by zero..
+            if denominator != 0:
+                estimated_pos = numerator / denominator
```

Comment:
Be aware that the result will be a double and not an integer! To use integer division, use //